hal/src/rtl872x/usart_hal.cpp

```diff
@@ -279,44 +322,59 @@ class Usart {
+            size_t alreadyCommitted = blockSize - rxBuffer_.acquirePending();
+            size_t toConsume = std::max(dmaAvailableInBuffer, transferredToDmaFromUart) - alreadyCommitted;
+            if (commit && toConsume > 0) {
+                if (transferredToDmaFromUart > dmaAvailableInBuffer) {
+                    uintptr_t expectedDstAddr = rxDmaInitStruct_.GDMA_DstAddr + transferredToDmaFromUart;
+                    if ((GDMA_BASE->CH[rxDmaInitStruct_.GDMA_ChNum].CFG_LOW & BIT_CFGX_LO_FIFO_EMPTY) == 0) {
+                        if (GDMA_GetDstAddr(rxDmaInitStruct_.GDMA_Index, rxDmaInitStruct_.GDMA_ChNum) < expectedDstAddr) {
+                            // Suspending DMA channel forces flushing of data into destination from GDMA FIFO
+                            GDMA_BASE->CH[rxDmaInitStruct_.GDMA_ChNum].CFG_LOW |= BIT_CFGX_LO_CH_SUSP;
+                            __DSB();
+                            __ISB();
+                            while (GDMA_GetDstAddr(rxDmaInitStruct_.GDMA_Index, rxDmaInitStruct_.GDMA_ChNum) < expectedDstAddr) {
+                                // XXX: spin around, this should be pretty fast
+                            }
+                            GDMA_BASE->CH[rxDmaInitStruct_.GDMA_ChNum].CFG_LOW &= ~(BIT_CFGX_LO_CH_SUSP);
+                            __DSB();
+                            __ISB();
+                        }
+                    }
+                }
+                rxBuffer_.acquireCommit(toConsume);
+                dcacheInvalidateAligned(rxDmaInitStruct_.GDMA_DstAddr + alreadyCommitted, toConsume);
+            }
+            return toConsume;
+        }
+        return 0;
+    }
+
     ssize_t data() {
         CHECK_TRUE(isEnabled(), SYSTEM_ERROR_INVALID_STATE);
-        auto uartInstance = uartTable_[index_].UARTx;
         ssize_t len = 0;
-        if (receiving_) {
-            RxLock lk(this);
-            if (!useInterrupt()) {
-                len = rxBuffer_.data();
-                const uint32_t curAddr = GDMA_GetDstAddr(rxDmaInitStruct_.GDMA_Index, rxDmaInitStruct_.GDMA_ChNum);
-                GDMA_Cmd(rxDmaInitStruct_.GDMA_Index, rxDmaInitStruct_.GDMA_ChNum, DISABLE);
-                const uint32_t receivedDma = curAddr - rxDmaInitStruct_.GDMA_DstAddr;
-                uint32_t inFifo = UART_ReceiveDataTO(uartInstance, (uint8_t*)curAddr, rxDmaInitStruct_.GDMA_BlockSize - receivedDma, 1);
-                uint32_t receivedTotal = receivedDma + inFifo;
-                // Clean Auto Reload Bit
-                GDMA_ChCleanAutoReload(rxDmaInitStruct_.GDMA_Index, rxDmaInitStruct_.GDMA_ChNum, CLEAN_RELOAD_DST);
-                // Clear Pending ISR
-                GDMA_ClearINT(rxDmaInitStruct_.GDMA_Index, rxDmaInitStruct_.GDMA_ChNum);
-                rxBuffer_.acquireCommit(receivedTotal, rxDmaInitStruct_.GDMA_BlockSize - receivedTotal/* cancel */);
-                len += receivedTotal;
-            } else {
```

Comment:
This code block is for interrupt mode.

Reply:
We don't normally transfer data out of internal peripheral FIFO in `data()`, so I'd argue we keep this as-is and let the interrupts put the data into the RX buffer and here just deal with whatever is already in it.

Reply:
I that case, should we call `UART_SetRxLevel(uartInstance, UART_RX_FIFOTRIG_LEVEL_1BYTES);` explicitly so that `Serial.available()` has value returned if there is data in FIFO but the ISR is not serviced.